### ISSUE TITLE
The event "firebase.messaging().onMessage" is not called on a real device

### ISSUE THREAD

**Issue by sturmenta** [invertase/react-native-firebase#1374]
### Issue

The event "firebase.messaging().OnMessage" is called in the simulator, but it is not called on the real device when a notification arrives

[notification arrives gif](https://drive.google.com/open?id=17M53YQY9sMlJx1PBNdAgCMCoCijP0iSw)

[event is not called gif](https://drive.google.com/open?id=1WKjq-bcudoLsT20P080PWPS2x86kP1U3)

[event is not called 2 gif](https://drive.google.com/open?id=13k74tbm83GGnmkQkr2lS4gG7wUJ-zbUn)

### Environment

1. Application Target Platform: Both

2. Development Operating System: macOS Sierra

3. `React Native` version: 0.56.0

4. `React Native Firebase` version: 4.3.8

5. `Firebase` Module: Cloud messaging

**Comment by nezaidu** [comment]
anyone? :(

**Comment by dackom** [comment]
I also have problem with this. Tried to put onMessage() listener in App.js and in some other components, but it doesnt trigger in foreground or background. What's more, I get this warning in console:

Sending `notifications_notification_received` with no listeners registered.

**Comment by amirasaran** [comment]
+1

**Comment by calintamasyopeso** [comment]
+1

**Comment by alexYopeso** [comment]
+1

**Comment by cristianplaias** [comment]
+1

**Comment by calintamasyopeso** [comment]
I have found the solution. Use the `onNotification` method as described below.

```javascript
firebase.notifications().onNotification((notification) => {
  firebase.notifications().displayNotification(notification);
});
```

**Comment by tvtoanit** [comment]
> I have found the solution. Use the `onNotification` method as described below.
> 
> ```js
> firebase.notifications().onNotification((notification) => {
>   firebase.notifications().displayNotification(notification);
> });
> ```

Can you help me? I want push notification code. I try send by messaging but not working. Bellow my code:
 ```
const message = new firebase.messaging.RemoteMessage()
            .setMessageId('fZiIctwDJCo:APA91bH5k3f06RorcsJUSwkrKEXDZW14pO4Vh-yWBA3cQt1grmyGNuG2MexoBwPl9J9LGwPtNP5dFq488mS5s7xf17Pt-SVMjtEVjqLedWMEXFyKr5Ldo1fmTEwf37e1eTnN0r17YUUU')
            .setTo('168153433661@gcm.googleapis.com')
            .setData(messageInput);
        await messaging.sendMessage(message);  
```

**Comment by aysfzaicmu** [comment]
@amirasaran i've been trying to do this however in both cases, the onMessage function is triggered and not onNotification..any idea where i might be going wrong

**Comment by muscaiu** [comment]
> 
> 
> Thanks @calintamasyopeso and @amirasaran.
> 
> In web working with this:
> 
> ```js
> firebase.messaging().onMessage(notification => {
> 	alert('Notification received!', notification);
> });
> ```
> 
> but in react-native working with this (after installing [firebase.notifications](https://rnfirebase.io/docs/v4.3.x/notifications/android))
> 
> ```js
> firebase.notifications().onNotification(notification => {
> 	alert('Notification received!', notification);
> });
> ```
> 
> with the same payload
> 
> ```json
> {
>     "to" : "",
>     "notification" : {}
> }
> ```

console.log and alert works also for me, but how i real notification doesn't.

**Comment by yardenst** [comment]
I'm using the web (NOT React Native)

version 7.3.0
Chrome 78

The callback I register to `firebase.messaging().onMessage` isn't called when tab in the foreground.

I tried to send a notification message and data message (neither works)

(I can confirm that setBackgroundMessageHandler does work)
(I can also confirm that when sending notification message and the tab is in the bg, it pops up)

Any idea? I need to handle the case when i receive a message and the tab is in the foreground.

**Comment by mikehardy** [comment]
@yardenst this is the repo for "**react-native**-firebase", I don't believe we can help

**Comment by yardenst** [comment]
@mikehardy Thanks (and sorry)

**Comment by VIRGO96** [comment]
> I'm using the web (NOT React Native)
> 
> version 7.3.0
> Chrome 78
> 
> The callback I register to `firebase.messaging().onMessage` isn't called when tab in the foreground.
> 
> I tried to send a notification message and data message (neither work)
> 
> (I can confirm that `setBackgroundMessageHandler` does work)
> (I can also confirm that when sending notification message and the tab is in the bg, it pops up)
> 
> Any idea? I need to handle the case when i receive a message and the tab is in the foreground.

did u get it to work i am having same problem in web

**Comment by yardenst** [comment]
Yes, I did get it to work (needed to update the firsebase version then it worked)

**Comment by VIRGO96** [comment]
@yardenst  what version were you using ?

**Comment by mikehardy** [comment]
@VIRGO96 in web? this is a react-native module. It does not do anything web-related.

**Comment by VIRGO96** [comment]
@mikehardy  yes i understand but @yardenst  is the only person i found to having the same problem as me and unfortunately he is on react native thread so thats why i commented here

**Comment by yardenst** [comment]
7.3.0

**Comment by kolyuchii** [comment]
Well, I've got the same issue.

React Native: 61
    "@react-native-firebase/app": "^8.2.0",
    "@react-native-firebase/auth": "^8.1.2",
    "@react-native-firebase/crashlytics": "^8.1.2",
    "@react-native-firebase/database": "^7.3.2",
    "@react-native-firebase/dynamic-links": "^7.3.2",
    "@react-native-firebase/firestore": "^7.4.3",
    "@react-native-firebase/functions": "^7.2.2",
    "@react-native-firebase/messaging": "^7.4.2",
    "@react-native-firebase/storage": "^7.2.2",

```
constructor(props) {
const authorizationStatus = await messaging().requestPermission();
                    if (
                        authorizationStatus ===
                            messaging.AuthorizationStatus.AUTHORIZED ||
                        authorizationStatus ===
                            messaging.AuthorizationStatus.PROVISIONAL
                    ) {
                        messaging()
                            .getToken()
                            .then((token) => {
                                console.log('token', token);
                                return UsersModel.updateCurrentUserData({
                                    token,
                                });
                            })
                            .catch((error) => console.log('App_:79', error));
                    } else {
                        console.log(
                            'User has notification permissions disabled',
                        );
                    }
}
componentDidMount() {
this.mesaging = messaging().onMessage((remoteMessage) => {
            console.log('Message', remoteMessage);
});
}
```

**Comment by Doetheman** [comment]
I am having this issue in flutter when sending custom data

**Comment by mikehardy** [comment]
@Doetheman - this is react-native not flutter, it will have little bearing on your problem, you should open an issue in FlutterFire package
